### PR TITLE
ALTAPPS-1129: Shared show paywall every third session

### PR DESCRIPTION
**Pull request by XanderZhu**
**YouTrack Issues**:
[ALTAPPS-1129](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-1129)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
- [x] All checks have been passed;
- [x] Changes have been checked locally.

**Description**
- Count app shows count in AppFeature to detect every 3rd show time;
- Fetch the subscription on the app startup and store it type in the app feature, to show a paywall only for freemium users;
- Subscribe on subscription changes to store fresh subscription type in the app feature;
- Implement `CurrentSubscriptionStateHolder` to store the subscription in the cache.

### CODE REVIEW

**Comment by XanderZhu** [comment]
@ivan-magda 
I've rewritten the code so the Paywall is shown every third time the app is opened from the background, including app startup.

File: shared/src/commonMain/kotlin/org/hyperskill/app/main/presentation/AppFeature.kt [hunk 1]
```diff
@@ -25,8 +31,15 @@ interface AppFeature {
             val isAuthorized: Boolean,
             val isMobileLeaderboardsEnabled: Boolean,
             internal val streakRecoveryState: StreakRecoveryFeature.State = StreakRecoveryFeature.State(),
-            internal val welcomeOnboardingState: WelcomeOnboardingFeature.State = WelcomeOnboardingFeature.State()
-        ) : State
+            internal val welcomeOnboardingState: WelcomeOnboardingFeature.State = WelcomeOnboardingFeature.State(),
+            internal val subscriptionType: SubscriptionType? = null,
+            internal val appShowsCount: Int = 1
+        ) : State {
+            internal val shouldShowPaywall: Boolean
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/main/presentation/AppFeature.kt:38]: I don't see the need for a `shouldShowPaywall` extension - we are using it only in the `AppReducer` let's move `APP_SHOWS_COUNT_TILL_PAYWALL` and `shouldShowPaywall` to reducer.

_For me feature is just the interface of the feature data flow, it should not contain business logic._

File: shared/src/commonMain/kotlin/org/hyperskill/app/main/presentation/AppFeature.kt [hunk 2]
```diff
@@ -35,11 +48,14 @@ interface AppFeature {
             val forceUpdate: Boolean = false
         ) : Message
 
-        data class UserAccountStatus(
+        object AppBecomesActive : Message
+
+        data class FetchAppStartupConfigSuccess(
             val profile: Profile,
+            val subscription: Subscription?,
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/main/presentation/AppFeature.kt:55]: according to the reducer usages we can pass `subscriptionType: SubscriptionType?`

File: shared/src/commonMain/kotlin/org/hyperskill/app/subscriptions/cache/CurrentSubscriptionStateHolderImpl.kt [hunk 3]
```diff
@@ -0,0 +1,33 @@
+package org.hyperskill.app.subscriptions.cache
+
+import com.russhwolf.settings.Settings
+import com.russhwolf.settings.contains
+import kotlinx.serialization.json.Json
+import org.hyperskill.app.subscriptions.data.source.CurrentSubscriptionStateHolder
+import org.hyperskill.app.subscriptions.domain.model.Subscription
+
+class CurrentSubscriptionStateHolderImpl(
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/subscriptions/cache/CurrentSubscriptionStateHolderImpl.kt:9]: internal

File: shared/src/commonMain/kotlin/org/hyperskill/app/subscriptions/cache/SubscriptionCacheKeys.kt [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package org.hyperskill.app.subscriptions.cache
+
+object SubscriptionCacheKeys {
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/subscriptions/cache/SubscriptionCacheKeys.kt:3]: internal

File: shared/src/commonMain/kotlin/org/hyperskill/app/subscriptions/domain/model/Subscription.kt [hunk 5]
```diff
@@ -3,8 +3,18 @@ package org.hyperskill.app.subscriptions.domain.model
 import kotlinx.datetime.Instant
 import kotlinx.serialization.SerialName
 import kotlinx.serialization.Serializable
+import org.hyperskill.app.subscriptions.cache.CurrentSubscriptionStateHolderImpl
 
 /**
+ * Represents a user profile.
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/subscriptions/domain/model/Subscription.kt:9]:
```suggestion
 * Represents a subscription.
```

**Review by ivan-magda — APPROVED** [review]

**Pull request merged**